torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp

```diff
@@ -39,6 +41,142 @@ c10::AliasAnalysisKind aliasAnalysisFromSchema() {
+    }
+  }
+  return last_use_node;
+}
+
+void InplaceMKLDNNSubgraph(std::shared_ptr<Graph> graph) {
+  // This function first calculates aliasing sets,
+  // then calculates the last node each aliasing set is alive for.
+  // Then we go through each node, if it's a node which has an equivalent
+  // inplace node and the aliasing set for its input is dead afer this node, we
+  // inplace it. Then we merge the aliasing sets for the input and output of the
+  // node and extend the liveness of the set. To inplace a node you need to
+  // prove device and dtype of the input and output are the same, which we've
+  // already done, and prove that the output size is the same as the input size,
+  // which is achieved by explicit Broadcast nodes (which we inserted for other
+  // reasons).
+
+  auto aliasDb = torch::make_unique<AliasDb>(graph);
+  std::unordered_map<Value*, ValueSetPtr> alias_mapping;
+  ValueSet set;
+  ValueSetPtr input_set = std::make_shared<ValueSet>(set);
+  for (Value* v : graph->inputs()) {
+    if (v->type()->cast<TensorType>()) {
+      input_set->insert(v);
+      alias_mapping[v] = input_set;
+    }
+  }
+
+  auto merge_sets = [&](Value* existing, Value* new_v) {
+    if (alias_mapping[existing] == alias_mapping[new_v]) {
+      return;
+    }
+    auto existing_set = alias_mapping[existing];
+    auto set_to_remove = alias_mapping[new_v];
+    for (auto it = set_to_remove->begin(); it != set_to_remove->end(); it++) {
+      existing_set->insert(*it);
+      alias_mapping[*it] = existing_set;
+    }
+  };
+
+  for (Node* n : graph->nodes()) {
+    for (Value* output : n->outputs()) {
+      if (!output->type()->cast<TensorType>()) {
+        continue;
+      }
+
+      std::unordered_set<Value*> new_set = {output};
+      alias_mapping[output] = std::make_shared<ValueSet>(new_set);
+      for (Value* input : n->inputs()) {
+        if (aliasDb->mayAlias(input, output)) {
+          merge_sets(input, output);
+        }
+      }
+    }
+  }
+
+  std::unordered_map<ValueSetPtr, Node*> set_liveness;
+  for (auto& set : alias_mapping) {
+    if (set_liveness.count(set.second)) {
+      continue;
+    }
+    Node* last = nullptr;
+    for (auto it = set.second->begin(); it != set.second->end(); it++) {
+      Value* v = *it;
+      auto k = v->node()->kind();
+      if (k == prim::Constant || k == prim::ConstantMKLDNNTensor ||
+          k == prim::Param) {
+        last = graph->return_node();
+        continue;
+      }
+
+      auto last_use = getLastUse(v);
+      if (!last || last_use->isAfter(last)) {
+        last = last_use;
+      }
+    }
+    set_liveness[set.second] = last;
+  }
+
+  std::vector<Node*> nodes_to_inplace;
+  auto add_to_inplace_set = [&](Node* node) {
+    // defer making the inplacing change because that would invalidate the old
+    // Node output Value*
+    nodes_to_inplace.push_back(node);
+    auto output_liveness_end =
+        set_liveness[alias_mapping[node->outputs().at(0)]];
```

Comment:
are we expecting multi-output nodes? should we assert here? and maybe a quick comment that we -re-arrange nodes to make `inputs().at(0)` reusable.

Reply:
there are no multi-output nodes, i can assert